File: src/lib/missions/handlers/ExpeditionMission.ts
/**
 * ExpeditionMission Handler
 *
 * Handles expedition missions to position 16 (deep space).
 * Expeditions have random outcomes including:
 * - Finding resources, ships, or dark matter
 * - Combat encounters with pirates or aliens
 * - Navigation events (delays or shortcuts)
 * - Catastrophic black hole events
 *
 * Mission type: 15
 */

import type { MissionType } from '@/types/database'
import { BaseMission } from '../BaseMission'
import {
  MissionContext,
  MissionArrivalResult,
  MissionReturnResult,
  MissionUpdate,
  Resources,
  ShipCounts,
  emptyResources,
  emptyShipCounts,
  getTotalShips,
  SHIP_KEYS,
} from '../types'
import {
  ExpeditionEventType,
  ExpeditionEvent,
  ExpeditionResult,
  EXPEDITION_PROBABILITIES,
  EXPEDITION_MESSAGES,
  FLEET_POWER_TIERS,
} from '@/lib/expeditions/types'

export class ExpeditionMission extends BaseMission {
  readonly missionType: MissionType = 'expedition'
  readonly hasReturn: boolean = true
  readonly name: string = 'Expedition'

  /**
   * Process expedition mission arrival
   *
   * 1. Roll for random event
   * 2. Apply event effects
   * 3. Create return mission with modified cargo/fleet/timing
   */
  async processArrival(context: MissionContext): Promise<MissionArrivalResult> {
    const { mission } = context

    const ships = this.getShips(mission)
    const totalShips = getTotalShips(ships)

    if (totalShips <= 0) {
      return this.errorArrival('No ships in fleet')
    }

    // Roll for event
    const event = this.rollExpeditionEvent()

    // Apply event effects
    const result = this.applyEvent(event, ships)

    // Calculate remaining ships after potential losses
    const remainingShips = this.subtractShips(ships, result.shipsLost)

    // Add any gained ships
    const finalShips = this.addShips(remainingShips, result.shipsGained)

    // Format coordinates for message
    const targetCoords = this.formatCoords(
      mission.destination_galaxy,
      mission.destination_system,
      mission.destination_position
    )

    // Build expedition report message
    const reportBody = this.buildExpeditionReport(
      targetCoords,
      result,
      ships,
      finalShips
    )

    const message = this.createMessage(
      mission.user_id,
      'expedition',
      'Expedition Report',
      reportBody
    )

    // Update user's dark matter if gained
    if (result.darkMatterGained > 0) {
      await this.addDarkMatter(mission.user_id, result.darkMatterGained)
    }

    // Handle time modifier for return trip
    const updates: MissionUpdate[] = []
    if (result.timeModifier !== 1.0) {
      // Time modifier is handled by creating return mission with modified timing
      // This will be processed by MissionProcessor
    }

    // If all ships were lost (black hole), no return
    if (getTotalShips(finalShips) === 0) {
      return this.successArrival(
        false, // No return
        emptyResources(),
        emptyShipCounts(),
        [message],
        updates
      )
    }

    // Return with resources, ships, and potentially modified timing
    return {
      success: true,
      shouldReturn: true,
      returnResources: result.resourcesGained,
      returnShips: finalShips,
      messages: [message],
      updates,
      // Custom field for time modifier (handled by processor)
      timeModifier: result.timeModifier,
    } as MissionArrivalResult & { timeModifier?: number }
  }

  /**
   * Process expedition return
   *
   * 1. Add resources to origin planet
   * 2. Add ships back to origin planet
   * 3. Send return message
   */
  async processReturn(context: MissionContext): Promise<MissionReturnResult> {
    const { mission, originPlanet } = context

    if (!originPlanet) {
      return this.errorReturn('Origin planet not found')
    }

    const ships = this.getShips(mission)
    const resources = this.getResources(mission)

    // Add ships back to origin planet
    const addShipsResult = await this.addShipsToPlanet(originPlanet.id, ships)
    if (!addShipsResult.success) {
      return this.errorReturn(`Failed to return ships: ${addShipsResult.error}`)
    }

    // Add resources to origin planet
    if (resources.metal > 0 || resources.crystal > 0 || resources.deuterium > 0) {
      const addResourcesResult = await this.addResourcesToPlanet(originPlanet.id, resources)
      if (!addResourcesResult.success) {
        return this.errorReturn(`Failed to deliver resources: ${addResourcesResult.error}`)
      }
    }

    // Format coordinates
    const originCoords = this.formatCoords(
      mission.destination_galaxy,
      mission.destination_system,
      mission.destination_position
    )

    const message = this.createMessage(
      mission.user_id,
      'expedition',
      'Expedition Fleet Returned',
      `Your expedition fleet has returned to ${originCoords}.\n\n` +
        `Resources delivered:\n` +
        `Metal: ${resources.metal.toLocaleString()}\n` +
        `Crystal: ${resources.crystal.toLocaleString()}\n` +
        `Deuterium: ${resources.deuterium.toLocaleString()}`
    )

    return this.successReturn([message], [])
  }

  /**
   * Roll for a random expedition event based on probabilities
   */
  private rollExpeditionEvent(): ExpeditionEvent {
    const roll = Math.random() * 100
    let cumulative = 0

    for (const [eventType, probability] of Object.entries(EXPEDITION_PROBABILITIES)) {
      cumulative += probability
      if (roll < cumulative) {
        const type = eventType as ExpeditionEventType
        const messages = EXPEDITION_MESSAGES[type]
        const message = messages[Math.floor(Math.random() * messages.length)]

        return {
          type,
          message,
        }
      }
    }

    // Fallback to nothing
    return {
      type: ExpeditionEventType.NOTHING,
      message: EXPEDITION_MESSAGES[ExpeditionEventType.NOTHING][0],
    }
  }

  /**
   * Apply event effects and calculate results
   */
  private applyEvent(event: ExpeditionEvent, fleet: ShipCounts): ExpeditionResult {
    const fleetPower = this.calculateFleetPower(fleet)
    const powerTier = this.getFleetPowerTier(fleetPower)

    const result: ExpeditionResult = {
      event,
      resourcesGained: emptyResources(),
      darkMatterGained: 0,
      shipsGained: {},
      shipsLost: {},
      timeModifier: 1.0,
    }

    switch (event.type) {
      case ExpeditionEventType.NOTHING:
        // No effect
        break

      case ExpeditionEventType.RESOURCES:
        result.resourcesGained = this.generateResourceReward(powerTier)
        event.resources = result.resourcesGained
        break

      case ExpeditionEventType.SHIPS:
        result.shipsGained = this.generateShipReward(powerTier)
        event.ships = result.shipsGained
        break

      case ExpeditionEventType.DARK_MATTER:
        result.darkMatterGained = this.generateDarkMatterReward(powerTier)
        event.value = result.darkMatterGained
        break

      case ExpeditionEventType.PIRATES:
        result.shipsLost = this.simulatePirateCombat(fleet, powerTier)
        event.ships = result.shipsLost
        break

      case ExpeditionEventType.ALIENS:
        result.shipsLost = this.simulateAlienCombat(fleet, powerTier)
        event.ships = result.shipsLost
        break

      case ExpeditionEventType.BLACK_HOLE:
        result.shipsLost = this.calculateBlackHoleLosses(fleet)
        event.ships = result.shipsLost
        break

      case ExpeditionEventType.DELAY:
        result.timeModifier = 1.5 // 50% slower return
        event.value = 50 // 50% delay
        break

      case ExpeditionEventType.SPEED_BONUS:
        result.timeModifier = 0.5 // 50% faster return
        event.value = 50 // 50% faster
        break
    }

    return result
  }

  /**
   * Calculate fleet power for reward scaling
   */
  private calculateFleetPower(fleet: ShipCounts): number {
    // Simplified power calculation based on ship values
    const shipPower: Record<keyof ShipCounts, number> = {
      light_fighter: 100,
      heavy_fighter: 300,
      cruiser: 2700,
      battleship: 6000,
      battlecruiser: 7000,
      bomber: 7500,
      destroyer: 11000,
      deathstar: 900000,
      small_cargo: 50,
      large_cargo: 150,
      colony_ship: 500,
      recycler: 200,
      espionage_probe: 1,
      reaper: 10000,
      pathfinder: 3000,
    }

    return SHIP_KEYS.reduce((total, key) => {
      return total + fleet[key] * (shipPower[key] || 100)
    }, 0)
  }

  /**
   * Get fleet power tier for scaling rewards
   */
  private getFleetPowerTier(power: number): number {
    if (power >= FLEET_POWER_TIERS.HUGE) return 5
    if (power >= FLEET_POWER_TIERS.LARGE) return 4
    if (power >= FLEET_POWER_TIERS.MEDIUM) return 3
    if (power >= FLEET_POWER_TIERS.SMALL) return 2
    return 1
  }

  /**
   * Generate resource reward based on fleet tier
   */
  private generateResourceReward(tier: number): Resources {
    const baseAmount = 10000 * tier
    const variance = 0.5 + Math.random() // 0.5 to 1.5

    return {
      metal: Math.floor(baseAmount * variance * (0.8 + Math.random() * 0.4)),
      crystal: Math.floor(baseAmount * variance * (0.6 + Math.random() * 0.4)),
      deuterium: Math.floor(baseAmount * variance * (0.3 + Math.random() * 0.2)),
    }
  }

  /**
   * Generate ship reward based on fleet tier
   */
  private generateShipReward(tier: number): Partial<ShipCounts> {
    const ships: Partial<ShipCounts> = {}

    // Possible ships to find (weighted by rarity)
    const possibleShips: Array<{ key: keyof ShipCounts; rarity: number }> = [
      { key: 'small_cargo', rarity: 30 },
      { key: 'large_cargo', rarity: 20 },
      { key: 'light_fighter', rarity: 25 },
      { key: 'heavy_fighter', rarity: 15 },
      { key: 'cruiser', rarity: 10 },
      { key: 'battleship', rarity: 5 },
      { key: 'espionage_probe', rarity: 25 },
      { key: 'recycler', rarity: 10 },
    ]

    // Roll for each ship type
    for (const ship of possibleShips) {
      const roll = Math.random() * 100
      if (roll < ship.rarity) {
        const baseAmount = Math.ceil(tier * (1 + Math.random() * 2))
        ships[ship.key] = baseAmount
      }
    }

    // Ensure at least one ship is found
    if (Object.keys(ships).length === 0) {
      ships.small_cargo = Math.ceil(tier * (1 + Math.random()))
    }

    return ships
  }

  /**
   * Generate dark matter reward
   */
  private generateDarkMatterReward(tier: number): number {
    const base = 50 * tier
    return Math.floor(base * (0.5 + Math.random()))
  }

  /**
   * Simulate pirate combat - generally weak enemies
   */
  private simulatePirateCombat(fleet: ShipCounts, tier: number): Partial<ShipCounts> {
    const losses: Partial<ShipCounts> = {}

    // Pirates are relatively weak - low losses
    const lossPercentage = Math.max(0.01, 0.1 - tier * 0.02)

    for (const key of SHIP_KEYS) {
      if (fleet[key] > 0) {
        const roll = Math.random()
        if (roll < lossPercentage) {
          losses[key] = Math.max(1, Math.floor(fleet[key] * lossPercentage * Math.random()))
        }
      }
    }

    return losses
  }

  /**
   * Simulate alien combat - stronger enemies
   */
  private simulateAlienCombat(fleet: ShipCounts, tier: number): Partial<ShipCounts> {
    const losses: Partial<ShipCounts> = {}

    // Aliens are stronger - moderate losses
    const lossPercentage = Math.max(0.05, 0.2 - tier * 0.03)

    for (const key of SHIP_KEYS) {
      if (fleet[key] > 0) {
        const roll = Math.random()
        if (roll < lossPercentage * 2) {
          losses[key] = Math.max(1, Math.floor(fleet[key] * lossPercentage * Math.random()))
        }
      }
    }

    return losses
  }

  /**
   * Calculate black hole losses - catastrophic
   */
  private calculateBlackHoleLosses(fleet: ShipCounts): Partial<ShipCounts> {
    const losses: Partial<ShipCounts> = {}

    // Black holes are devastating - 30-100% of fleet lost
    const totalLossPercentage = 0.3 + Math.random() * 0.7

    for (const key of SHIP_KEYS) {
      if (fleet[key] > 0) {
        losses[key] = Math.floor(fleet[key] * totalLossPercentage)
      }
    }

    return losses
  }

  /**
   * Subtract ships from fleet
   */
  private subtractShips(fleet: ShipCounts, losses: Partial<ShipCounts>): ShipCounts {
    const result = { ...fleet }
    for (const key of SHIP_KEYS) {
      if (losses[key]) {
        result[key] = Math.max(0, result[key] - losses[key])
      }
    }
    return result
  }

  /**
   * Add ships to fleet
   */
  private addShips(fleet: ShipCounts, gained: Partial<ShipCounts>): ShipCounts {
    const result = { ...fleet }
    for (const key of SHIP_KEYS) {
      if (gained[key]) {
        result[key] = result[key] + gained[key]
      }
    }
    return result
  }

  /**
   * Add dark matter to user
   */
  private async addDarkMatter(userId: string, amount: number): Promise<void> {
    const { data: user } = await this.supabase
      .from('users')
      .select('dark_matter')
      .eq('id', userId)
      .single()

    if (user) {
      await this.supabase
        .from('users')
        .update({
          dark_matter: user.dark_matter + amount,
        })
        .eq('id', userId)
    }
  }

  /**
   * Build expedition report message
   */
  private buildExpeditionReport(
    coords: string,
    result: ExpeditionResult,
    originalFleet: ShipCounts,
    finalFleet: ShipCounts
  ): string {
    const lines: string[] = [
      `Expedition Report from ${coords}`,
      '',
      result.event.message,
      '',
    ]

    // Add resource gains
    if (result.resourcesGained.metal > 0 || result.resourcesGained.crystal > 0 || result.resourcesGained.deuterium > 0) {
      lines.push('Resources found:')
      if (result.resourcesGained.metal > 0) {
        lines.push(`  Metal: ${result.resourcesGained.metal.toLocaleString()}`)
      }
      if (result.resourcesGained.crystal > 0) {
        lines.push(`  Crystal: ${result.resourcesGained.crystal.toLocaleString()}`)
      }
      if (result.resourcesGained.deuterium > 0) {
        lines.push(`  Deuterium: ${result.resourcesGained.deuterium.toLocaleString()}`)
      }
      lines.push('')
    }

    // Add dark matter gains
    if (result.darkMatterGained > 0) {
      lines.push(`Dark Matter found: ${result.darkMatterGained.toLocaleString()}`)
      lines.push('')
    }

    // Add ship gains
    const gainedShips = Object.entries(result.shipsGained).filter(([, v]) => v && v > 0)
    if (gainedShips.length > 0) {
      lines.push('Ships found:')
      for (const [ship, count] of gainedShips) {
        lines.push(`  ${this.formatShipName(ship)}: ${count}`)
      }
      lines.push('')
    }

    // Add ship losses
    const lostShips = Object.entries(result.shipsLost).filter(([, v]) => v && v > 0)
    if (lostShips.length > 0) {
      lines.push('Ships lost:')
      for (const [ship, count] of lostShips) {
        lines.push(`  ${this.formatShipName(ship)}: ${count}`)
      }
      lines.push('')
    }

    // Add time modifier info
    if (result.timeModifier !== 1.0) {
      if (result.timeModifier < 1.0) {
        lines.push(`Return time reduced by ${Math.round((1 - result.timeModifier) * 100)}%!`)
      } else {
        lines.push(`Return time increased by ${Math.round((result.timeModifier - 1) * 100)}%.`)
      }
      lines.push('')
    }

    // Fleet status
    const totalRemaining = getTotalShips(finalFleet)
    if (totalRemaining === 0) {
      lines.push('The entire fleet was lost.')
    } else {
      lines.push('The fleet is returning to base.')
    }

    return lines.join('\n')
  }

  /**
   * Format ship key to readable name
   */
  private formatShipName(key: string): string {
    return key
      .split('_')
      .map(word => word.charAt(0).toUpperCase() + word.slice(1))
      .join(' ')
  }
}

File: unity_client/Assets/Scripts/SharedLib/Missions/Handlers/ExpeditionMission.cs
// =============================================================================
// ExpeditionMission.cs — Expedition mission handler
// Port of: src/lib/missions/handlers/ExpeditionMission.ts
// Namespace: OGameX.SharedLib.Missions.Handlers
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace OGameX.SharedLib.Missions.Handlers
{
    /// <summary>
    /// Handles expedition missions to deep space (position 16).
    /// 
    /// Expeditions produce random events with weighted probabilities.
    /// Fleet power determines reward tiers.
    /// 
    /// Possible outcomes:
    /// - Nothing (30%) — no event
    /// - Resources found (25%) — metal/crystal/deuterium
    /// - Dark matter (10%) — premium currency
    /// - Ships found (12%) — free ships discovered
    /// - Pirate attack (8%) — combat vs. NPC pirates
    /// - Alien attack (5%) — combat vs. NPC aliens
    /// - Delay (5%) — fleet delayed
    /// - Early return (3%) — fleet returns early
    /// - Black hole (2%) — fleet partially/fully destroyed
    /// 
    /// Reward scaling is based on fleet power tiers:
    /// Tier 1: 0-5k, Tier 2: 5k-25k, Tier 3: 25k-100k,
    /// Tier 4: 100k-500k, Tier 5: 500k+
    /// </summary>
    public class ExpeditionMission : BaseMission
    {
        public override MissionType HandlerMissionType => MissionType.Expedition;
        public override bool HasReturn => true;
        public override string Name => "Expedition";

        // =====================================================================
        // OUTCOME PROBABILITY TABLE
        // =====================================================================

        /// <summary>Weighted outcome table (probabilities sum to 1.0).</summary>
        private static readonly (ExpeditionOutcome Outcome, double Weight)[] OutcomeTable =
        {
            (ExpeditionOutcome.Nothing, 0.30),
            (ExpeditionOutcome.ResourcesFound, 0.25),
            (ExpeditionOutcome.DarkMatter, 0.10),
            (ExpeditionOutcome.ShipFound, 0.12),
            (ExpeditionOutcome.PirateAttack, 0.08),
            (ExpeditionOutcome.AlienAttack, 0.05),
            (ExpeditionOutcome.Delay, 0.05),
            (ExpeditionOutcome.EarlyReturn, 0.03),
            (ExpeditionOutcome.BlackHole, 0.02),
        };

        // =====================================================================
        // FLEET POWER TIERS
        // =====================================================================

        /// <summary>Fleet power thresholds for reward tiers.</summary>
        private static readonly (int Min, int Max, double Multiplier)[] FleetPowerTiers =
        {
            (0, 5000, 1.0),
            (5000, 25000, 2.0),
            (25000, 100000, 4.0),
            (100000, 500000, 8.0),
            (500000, int.MaxValue, 15.0),
        };

        // =====================================================================
        // RESOURCE REWARDS BASE VALUES
        // =====================================================================

        private const double BaseMetalReward = 10000;
        private const double BaseCrystalReward = 5000;
        private const double BaseDeuteriumReward = 2000;
        private const int BaseDarkMatterReward = 50;

        // =====================================================================
        // ARRIVAL PROCESSING
        // =====================================================================

        public override MissionArrivalResult ProcessArrival(MissionContext context)
        {
            var mission = context.Mission;
            var ships = GetShips(mission);
            var resources = GetResources(mission);

            var coords = FormatCoords(
                mission.DestinationGalaxy,
                mission.DestinationSystem,
                mission.DestinationPosition);

            // Calculate fleet power
            int fleetPower = CalculateFleetPower(ships);
            double rewardMultiplier = GetRewardMultiplier(fleetPower);

            // Roll expedition outcome
            var outcome = RollOutcome();

            // Process outcome
            var eventResult = ProcessOutcome(outcome, ships, rewardMultiplier, coords);

            // Build messages
            var messages = new List<MissionMessage>();
            messages.Add(CreateMessage(
                mission.UserId,
                MessageType.Expedition,
                FormatOutcomeSubject(eventResult.Outcome),
                eventResult.EventMessage
            ));

            // Calculate return ships and resources
            var returnShips = ships.Clone();
            var returnResources = resources.Clone();

            // Apply gains
            if (eventResult.FoundResources != null)
            {
                returnResources.Add(eventResult.FoundResources);
            }

            if (eventResult.ShipsGained != null)
            {
                foreach (var kv in eventResult.ShipsGained)
                    returnShips.Add(kv.Key, kv.Value);
            }

            // Apply losses
            if (eventResult.ShipsLost != null)
            {
                foreach (var kv in eventResult.ShipsLost)
                {
                    int current = returnShips.Get(kv.Key);
                    returnShips.Set(kv.Key, Math.Max(0, current - kv.Value));
                }
            }

            bool shouldReturn = returnShips.GetTotal() > 0;

            return SuccessArrival(
                shouldReturn: shouldReturn,
                returnResources: returnResources,
                returnShips: returnShips,
                messages: messages
            );
        }

        // =====================================================================
        // RETURN PROCESSING
        // =====================================================================

        public override MissionReturnResult ProcessReturn(MissionContext context)
        {
            var mission = context.Mission;
            var resources = GetResources(mission);

            var originCoords = FormatCoords(
                mission.DestinationGalaxy,
                mission.DestinationSystem,
                mission.DestinationPosition);

            string messageBody = $"Your expedition fleet has returned to {originCoords}";
            if (resources.HasResources())
                messageBody += $" carrying:\n{FormatResources(resources)}";
            else
                messageBody += ".";

            var message = CreateMessage(
                mission.UserId,
                MessageType.Expedition,
                "Expedition Returned",
                messageBody
            );

            return SuccessReturn(new List<MissionMessage> { message });
        }

        // =====================================================================
        // FLEET POWER CALCULATION
        // =====================================================================

        /// <summary>
        /// Calculate fleet power based on ship counts.
        /// Each ship type has a different combat power value.
        /// </summary>
        public static int CalculateFleetPower(ShipCounts ships)
        {
            // Approximate combat values per ship type
            int power = 0;
            power += ships.LightFighter * 15;
            power += ships.HeavyFighter * 45;
            power += ships.Cruiser * 170;
            power += ships.Battleship * 395;
            power += ships.Battlecruiser * 510;
            power += ships.Bomber * 500;
            power += ships.Destroyer * 850;
            power += ships.Deathstar * 50000;
            power += ships.SmallCargo * 10;
            power += ships.LargeCargo * 15;
            power += ships.ColonyShip * 5;
            power += ships.Recycler * 8;
            power += ships.EspionageProbe * 1;
            power += ships.Reaper * 2800;
            power += ships.Pathfinder * 310;
            return power;
        }

        /// <summary>Get the reward multiplier for a given fleet power tier.</summary>
        private static double GetRewardMultiplier(int fleetPower)
        {
            foreach (var tier in FleetPowerTiers)
            {
                if (fleetPower >= tier.Min && fleetPower < tier.Max)
                    return tier.Multiplier;
            }
            return 1.0;
        }

        // =====================================================================
        // OUTCOME PROCESSING
        // =====================================================================

        /// <summary>Roll a random expedition outcome from the probability table.</summary>
        protected virtual ExpeditionOutcome RollOutcome()
        {
            var random = new Random();
            double roll = random.NextDouble();
            double cumulative = 0;

            foreach (var (outcome, weight) in OutcomeTable)
            {
                cumulative += weight;
                if (roll < cumulative)
                    return outcome;
            }

            return ExpeditionOutcome.Nothing;
        }

        /// <summary>Process an expedition outcome and generate the event result.</summary>
        private ExpeditionEventResult ProcessOutcome(
            ExpeditionOutcome outcome,
            ShipCounts ships,
            double multiplier,
            string coords)
        {
            var random = new Random();

            return outcome switch
            {
                ExpeditionOutcome.Nothing => new ExpeditionEventResult
                {
                    Outcome = outcome,
                    EventMessage = $"Your expedition at {coords} scouted the area but found nothing of interest. The fleet is returning.",
                },

                ExpeditionOutcome.ResourcesFound => ProcessResourceFind(multiplier, coords, random),
                ExpeditionOutcome.DarkMatter => ProcessDarkMatterFind(multiplier, coords, random),
                ExpeditionOutcome.ShipFound => ProcessShipFind(multiplier, coords, random),
                ExpeditionOutcome.PirateAttack => ProcessPirateAttack(ships, multiplier, coords, random),
                ExpeditionOutcome.AlienAttack => ProcessAlienAttack(ships, multiplier, coords, random),
                ExpeditionOutcome.Delay => ProcessDelay(coords, random),
                ExpeditionOutcome.EarlyReturn => ProcessEarlyReturn(coords),
                ExpeditionOutcome.BlackHole => ProcessBlackHole(ships, coords, random),
                _ => new ExpeditionEventResult
                {
                    Outcome = ExpeditionOutcome.Nothing,
                    EventMessage = "Nothing happened.",
                },
            };
        }

        // =====================================================================
        // EVENT PROCESSORS
        // =====================================================================

        private ExpeditionEventResult ProcessResourceFind(double multiplier, string coords, Random rng)
        {
            double metal = Math.Floor(BaseMetalReward * multiplier * (0.5 + rng.NextDouble()));
            double crystal = Math.Floor(BaseCrystalReward * multiplier * (0.5 + rng.NextDouble()));
            double deut = Math.Floor(BaseDeuteriumReward * multiplier * (0.5 + rng.NextDouble()));

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.ResourcesFound,
                FoundResources = new Resources(metal, crystal, deut),
                EventMessage = $"Your expedition at {coords} discovered an abandoned cargo container!\n\n" +
                               $"Resources found:\nMetal: {metal:N0}\nCrystal: {crystal:N0}\nDeuterium: {deut:N0}",
            };
        }

        private ExpeditionEventResult ProcessDarkMatterFind(double multiplier, string coords, Random rng)
        {
            int dm = (int)Math.Floor(BaseDarkMatterReward * multiplier * (0.5 + rng.NextDouble()));

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.DarkMatter,
                DarkMatterFound = dm,
                EventMessage = $"Your expedition at {coords} found a cache of Dark Matter!\n\nDark Matter found: {dm:N0}",
            };
        }

        private ExpeditionEventResult ProcessShipFind(double multiplier, string coords, Random rng)
        {
            // Choose a random ship type to find
            var findableShips = new[]
            {
                (ShipType.LightFighter, 5),
                (ShipType.HeavyFighter, 3),
                (ShipType.Cruiser, 2),
                (ShipType.SmallCargo, 4),
                (ShipType.LargeCargo, 2),
                (ShipType.EspionageProbe, 8),
                (ShipType.Pathfinder, 1),
            };

            var pick = findableShips[rng.Next(findableShips.Length)];
            int count = (int)Math.Max(1, Math.Floor(pick.Item2 * multiplier * (0.5 + rng.NextDouble())));

            var gained = new Dictionary<ShipType, int> { { pick.Item1, count } };

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.ShipFound,
                ShipsGained = gained,
                EventMessage = $"Your expedition at {coords} found abandoned ships drifting in space!\n\n" +
                               $"Ships found: {count}x {ShipCounts.DisplayNames[pick.Item1]}",
            };
        }

        private ExpeditionEventResult ProcessPirateAttack(ShipCounts ships, double multiplier, string coords, Random rng)
        {
            // Pirates cause random ship losses (5-20% of fleet)
            double lossRate = 0.05 + rng.NextDouble() * 0.15;
            var losses = CalculateFleetLosses(ships, lossRate);

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.PirateAttack,
                ShipsLost = losses,
                EventMessage = $"Your expedition at {coords} was ambushed by space pirates!\n\n" +
                               $"Ships lost in combat:\n{FormatShipLossDict(losses)}",
            };
        }

        private ExpeditionEventResult ProcessAlienAttack(ShipCounts ships, double multiplier, string coords, Random rng)
        {
            // Aliens are stronger — 10-35% losses
            double lossRate = 0.10 + rng.NextDouble() * 0.25;
            var losses = CalculateFleetLosses(ships, lossRate);

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.AlienAttack,
                ShipsLost = losses,
                EventMessage = $"Your expedition at {coords} encountered hostile alien forces!\n\n" +
                               $"Ships lost in combat:\n{FormatShipLossDict(losses)}",
            };
        }

        private ExpeditionEventResult ProcessDelay(string coords, Random rng)
        {
            double modifier = 1.0 + (0.5 + rng.NextDouble()); // 1.5x to 2.0x

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.Delay,
                TimeModifier = modifier,
                EventMessage = $"Your expedition at {coords} encountered a navigational anomaly " +
                               $"and will be delayed. Return time increased by {(modifier - 1.0) * 100:F0}%.",
            };
        }

        private ExpeditionEventResult ProcessEarlyReturn(string coords)
        {
            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.EarlyReturn,
                TimeModifier = 0.5,
                EventMessage = $"Your expedition at {coords} found a shortcut through a wormhole " +
                               $"and is returning early!",
            };
        }

        private ExpeditionEventResult ProcessBlackHole(ShipCounts ships, string coords, Random rng)
        {
            // Black hole: 30-100% fleet loss
            double lossRate = 0.30 + rng.NextDouble() * 0.70;
            var losses = CalculateFleetLosses(ships, lossRate);

            string severity = lossRate > 0.8 ? "catastrophic" : lossRate > 0.5 ? "severe" : "partial";

            return new ExpeditionEventResult
            {
                Outcome = ExpeditionOutcome.BlackHole,
                ShipsLost = losses,
                EventMessage = $"Your expedition at {coords} was caught in a gravitational anomaly!\n\n" +
                               $"The black hole caused {severity} damage to the fleet.\n" +
                               $"Ships lost:\n{FormatShipLossDict(losses)}",
            };
        }

        // =====================================================================
        // HELPERS
        // =====================================================================

        /// <summary>Calculate proportional fleet losses.</summary>
        private Dictionary<ShipType, int> CalculateFleetLosses(ShipCounts ships, double lossRate)
        {
            var losses = new Dictionary<ShipType, int>();
            foreach (var t in ShipCounts.AllTypes)
            {
                int count = ships.Get(t);
                if (count > 0)
                {
                    int lost = (int)Math.Ceiling(count * lossRate);
                    lost = Math.Min(lost, count);
                    if (lost > 0)
                        losses[t] = lost;
                }
            }
            return losses;
        }

        private string FormatShipLossDict(Dictionary<ShipType, int> losses)
        {
            var lines = losses
                .Where(kv => kv.Value > 0)
                .Select(kv => $"  {ShipCounts.DisplayNames[kv.Key]}: {kv.Value:N0}");
            return lines.Any() ? string.Join("\n", lines) : "  None";
        }

        private string FormatOutcomeSubject(ExpeditionOutcome outcome) => outcome switch
        {
            ExpeditionOutcome.Nothing => "Expedition - No Findings",
            ExpeditionOutcome.ResourcesFound => "Expedition - Resources Found!",
            ExpeditionOutcome.DarkMatter => "Expedition - Dark Matter Found!",
            ExpeditionOutcome.ShipFound => "Expedition - Ships Discovered!",
            ExpeditionOutcome.PirateAttack => "Expedition - Pirate Attack!",
            ExpeditionOutcome.AlienAttack => "Expedition - Alien Encounter!",
            ExpeditionOutcome.Delay => "Expedition - Navigation Delay",
            ExpeditionOutcome.EarlyReturn => "Expedition - Early Return",
            ExpeditionOutcome.BlackHole => "Expedition - Black Hole!",
            _ => "Expedition Report",
        };
    }
}
